src/input/sources/ytdl.rs

```diff
@@ -56,6 +56,52 @@ impl YoutubeDl {
         }
     }
 
+    /// Creates a lazy(?) request to search youtube for a videos matching `query`.
+    ///
+    /// [`new`]: Self::new
+    #[must_use]
+    pub fn new_yt_search(client: Client, query: String) -> Self {
+        Self::new_ytdl_like(YOUTUBE_DL_COMMAND, client, format!("ytsearch5:{}", query))
+    }
+
+    /// Does a search query for the given url, returning a list of possible matches
+    /// which are youtube urls.
+    pub async fn search_query(&mut self) -> Result<Vec<String>, AudioStreamError> {
+        let search_str = if self.url.starts_with("ytsearch") {
+            self.url.clone()
+        } else {
+            format!("ytsearch5:{}", self.url)
+        };
+
+        let ytdl_args = [
+            "-s",
+            &search_str,
+            "--get-id",
+            "--get-title",
+            "--get-duration",
+        ];
+
+        let output = Command::new(self.program)
+            .args(ytdl_args)
+            .output()
+            .await
+            .map_err(|e| {
+                AudioStreamError::Fail(if e.kind() == ErrorKind::NotFound {
+                    format!("could not find executable '{}' on path", self.program).into()
+                } else {
+                    Box::new(e)
+                })
+            })?;
```

Comment:
While converting `query()` to support ndjson (`self.query_many(1).pop()`?), you could probably replace this function to:
* call an inner query function (`YoutubeDl::query_many(&mut self, n: usize)`?) to get a `Vec<AuxMetadata>`.
* You can then either extract `source_url` if you want a `Vec<String>` or just return all the metadata verbatim.

Reply:
I just returned it as AuxMetadata, lmk if the approach needs further change,